File: packages/python/src/agent_capture/schema/canonical.py
"""Canonical span serialization for hashing.

The provenance hash chain depends on Python and TypeScript producing
byte-identical bytes for a given logical span. This module owns the
canonical form. Any change here is a schema-breaking change and must bump
:data:`agent_capture.schema.provenance.SCHEMA_VERSION`.

Rules of canonicalization (mirror these in the TypeScript implementation):

1. Object keys are sorted lexicographically (UTF-8 codepoint order).
2. No insignificant whitespace. Separators are ``","`` and ``":"``.
3. UTF-8 encoding of the resulting JSON string is the byte input to SHA-256.
4. ``datetime`` values are rendered as RFC 3339 / ISO 8601 strings in UTC,
   nanosecond precision when available, ``Z`` suffix (no offset notation).
5. The ``provenance`` object is excluded from the hash input (a span cannot
   hash over its own hash).
6. ``None``/``null`` values are preserved (not stripped). The presence vs.
   absence of a key is significant.
7. Floats are emitted with Python's ``repr`` semantics ported to JSON:
   shortest round-trip decimal. NaN/Infinity are forbidden.

The output of :func:`canonical_bytes` is what gets hashed; the output of
:func:`canonical_json` is the same bytes decoded to ``str`` for debugging.
"""

from __future__ import annotations

import hashlib
import json
import math
from datetime import UTC, datetime
from typing import Any

from pydantic import BaseModel


def _normalize(value: Any) -> Any:
    """Recursively convert a value into JSON-canonicalizable primitives.

    Pydantic models are dumped via ``model_dump(mode='json')`` which already
    handles enums, datetimes, etc. — but we re-walk the result to guarantee
    every nested datetime/float is normalized consistently regardless of
    Pydantic's exact mode behavior.
    """
    if isinstance(value, BaseModel):
        return _normalize(value.model_dump(mode="json"))
    if isinstance(value, dict):
        return {str(k): _normalize(v) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [_normalize(v) for v in value]
    if isinstance(value, datetime):
        return _format_datetime(value)
    if isinstance(value, float):
        if math.isnan(value) or math.isinf(value):
            raise ValueError("Canonical form forbids NaN and Infinity.")
        return value
    return value


def _format_datetime(dt: datetime) -> str:
    """Render a datetime as a canonical RFC 3339 UTC string."""
    dt = dt.replace(tzinfo=UTC) if dt.tzinfo is None else dt.astimezone(UTC)
    # microsecond precision is the most Python natively offers; pad to a fixed
    # width so the serialized form is deterministic.
    base = dt.strftime("%Y-%m-%dT%H:%M:%S")
    micros = dt.microsecond
    return f"{base}.{micros:06d}Z"


def canonical_bytes(span_like: BaseModel | dict[str, Any], *, exclude_provenance: bool = True) -> bytes:
    """Return the canonical UTF-8 byte string for hashing.

    Args:
        span_like: A :class:`~agent_capture.schema.span.Span`, any Pydantic
            model, or a plain dict produced from one.
        exclude_provenance: When ``True`` (the default), the ``provenance``
            key is removed before serialization. This is the correct setting
            when computing a span's own ``content_hash``. Pass ``False`` for
            external verification once a chain already includes a hash.

    Returns:
        UTF-8 bytes ready to feed into ``hashlib.sha256(...)``.
    """
    payload = span_like.model_dump(mode="json") if isinstance(span_like, BaseModel) else dict(span_like)

    if exclude_provenance and isinstance(payload, dict):
        payload = {k: v for k, v in payload.items() if k != "provenance"}

    normalized = _normalize(payload)
    return json.dumps(
        normalized,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
        allow_nan=False,
    ).encode("utf-8")


def canonical_json(span_like: BaseModel | dict[str, Any], *, exclude_provenance: bool = True) -> str:
    """String form of :func:`canonical_bytes`. For debugging / golden fixtures."""
    return canonical_bytes(span_like, exclude_provenance=exclude_provenance).decode("utf-8")


def content_hash(span_like: BaseModel | dict[str, Any]) -> str:
    """Compute the SHA-256 ``content_hash`` for a span.

    Excludes the ``provenance`` key from the input — see module docstring.
    """
    return hashlib.sha256(canonical_bytes(span_like, exclude_provenance=True)).hexdigest()

File: packages/typescript/src/schema/canonical.ts
/**
 * Canonical span serialization for hashing.
 *
 * Must produce byte-identical output to the Python implementation at
 * packages/python/src/agent_capture/schema/canonical.py for any given
 * logical span. The provenance hash chain depends on this. Any change
 * here is a schema-breaking change and must bump SCHEMA_VERSION in
 * lockstep with the Python side.
 *
 * Rules:
 *
 * 1. Object keys sorted lexicographically (UTF-8 codepoint order).
 * 2. No insignificant whitespace; separators are "," and ":".
 * 3. UTF-8 encoding of the JSON string is the byte input to SHA-256.
 * 4. Date / datetime values rendered as
 *    "YYYY-MM-DDTHH:MM:SS.ffffffZ" (microsecond precision, UTC, "Z").
 *    JavaScript's Date has millisecond precision; the last three
 *    microsecond digits are emitted as "000" so the output exactly
 *    matches a Python datetime with millis truncated. A pre-formatted
 *    string matching the pattern is passed through verbatim.
 * 5. The "provenance" key is excluded from the hash input by default
 *    (a span cannot hash over its own hash).
 * 6. null values preserved — presence vs. absence of a key matters.
 * 7. Floats: standard JSON.stringify shortest round-trip. NaN/Infinity
 *    are forbidden.
 */

import { createHash } from "node:crypto";

const DATETIME_RE =
  /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}Z$/;

export interface CanonicalOptions {
  /** When true (default), strip the `provenance` key before serialization. */
  excludeProvenance?: boolean;
}

/**
 * Recursively normalize a value into JSON-canonicalizable primitives.
 * Anything not handled here passes through to JSON.stringify, which
 * matches Python's json module for primitive numbers and strings.
 */
function normalize(value: unknown): unknown {
  if (value === null || value === undefined) return value;
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new Error("Canonical form forbids NaN and Infinity.");
    }
    return value;
  }
  if (typeof value === "boolean" || typeof value === "string") {
    return value;
  }
  if (value instanceof Date) {
    return formatDate(value);
  }
  if (Array.isArray(value)) {
    return value.map(normalize);
  }
  if (typeof value === "object") {
    const out: Record<string, unknown> = {};
    const keys = Object.keys(value as Record<string, unknown>).sort();
    for (const k of keys) {
      out[k] = normalize((value as Record<string, unknown>)[k]);
    }
    return out;
  }
  return value;
}

function formatDate(d: Date): string {
  // YYYY-MM-DDTHH:MM:SS.ffffffZ — pad millis (3 digits) with "000".
  const pad = (n: number, width: number) => String(n).padStart(width, "0");
  const yyyy = pad(d.getUTCFullYear(), 4);
  const mm = pad(d.getUTCMonth() + 1, 2);
  const dd = pad(d.getUTCDate(), 2);
  const hh = pad(d.getUTCHours(), 2);
  const mi = pad(d.getUTCMinutes(), 2);
  const ss = pad(d.getUTCSeconds(), 2);
  const ms = pad(d.getUTCMilliseconds(), 3);
  return `${yyyy}-${mm}-${dd}T${hh}:${mi}:${ss}.${ms}000Z`;
}

/**
 * Stable JSON.stringify with sorted keys, no whitespace, no NaN.
 * Used by `canonicalBytes`. Operates on already-normalized input.
 */
function stableStringify(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new Error("Canonical form forbids NaN and Infinity.");
    }
    // JSON.stringify produces shortest round-trip for finite numbers.
    return JSON.stringify(value);
  }
  if (typeof value === "boolean" || typeof value === "string") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(stableStringify).join(",") + "]";
  }
  if (typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>);
    // keys are already sorted by `normalize`, but guard anyway.
    entries.sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
    return (
      "{" +
      entries.map(([k, v]) => JSON.stringify(k) + ":" + stableStringify(v)).join(",") +
      "}"
    );
  }
  throw new Error(`Cannot canonicalize value of type ${typeof value}`);
}

/**
 * Validate that any datetime-shaped string in the input matches the
 * canonical pattern. Used as a guard so accidental ISO 8601 strings
 * with timezone offsets or wrong precision don't silently produce
 * the wrong hash.
 */
function assertDatetimeShapes(value: unknown): void {
  if (typeof value === "string") {
    // Heuristic: if it looks like an ISO date but doesn't match canonical
    // form, complain.
    if (
      /^\d{4}-\d{2}-\d{2}T/.test(value) &&
      !DATETIME_RE.test(value)
    ) {
      throw new Error(
        `Datetime string ${JSON.stringify(value)} is not in canonical form ` +
          `(YYYY-MM-DDTHH:MM:SS.ffffffZ). Pass a Date or pre-format manually.`,
      );
    }
    return;
  }
  if (Array.isArray(value)) {
    for (const v of value) assertDatetimeShapes(v);
    return;
  }
  if (value && typeof value === "object") {
    for (const v of Object.values(value as Record<string, unknown>)) {
      assertDatetimeShapes(v);
    }
  }
}

/**
 * Return the canonical UTF-8 byte string for `value`.
 *
 * @param value     Any JSON-shaped object. Pydantic-dumped span objects work.
 * @param options.excludeProvenance  When true (default), strip the top-level
 *                  `provenance` key before serialization.
 */
export function canonicalBytes(
  value: unknown,
  options: CanonicalOptions = {},
): Uint8Array {
  const { excludeProvenance = true } = options;
  let working: unknown = value;
  if (
    excludeProvenance &&
    working &&
    typeof working === "object" &&
    !Array.isArray(working) &&
    "provenance" in (working as Record<string, unknown>)
  ) {
    const { provenance: _drop, ...rest } = working as Record<string, unknown>;
    working = rest;
  }
  assertDatetimeShapes(working);
  const normalized = normalize(working);
  const json = stableStringify(normalized);
  return new TextEncoder().encode(json);
}

/** String form of `canonicalBytes`; useful for debugging and golden fixtures. */
export function canonicalJson(
  value: unknown,
  options: CanonicalOptions = {},
): string {
  return new TextDecoder("utf-8").decode(canonicalBytes(value, options));
}

/** Hex-encoded SHA-256 of the canonical form. */
export function contentHash(value: unknown): string {
  const bytes = canonicalBytes(value, { excludeProvenance: true });
  return createHash("sha256").update(bytes).digest("hex");
}
